spmimage/linear_model/admm.py

```diff
@@ -253,3 +254,21 @@ def generate_transform_matrix(self, n_features: int) -> np.ndarray:
         if self.diagonal:
             return sp.sparse.dia_matrix(generated)
         return generated
+
+
+class TrendFilteringADMM(GeneralizedLasso):
+    def __init__(self, margin=1, alpha=1.0, rho=1.0, fit_intercept=True,
+                 normalize=False, copy_X=True, max_iter=1000,
+                 tol=1e-4):
+        super().__init__(alpha=alpha, rho=rho, fit_intercept=fit_intercept,
+                         normalize=normalize, copy_X=copy_X, max_iter=max_iter,
+                         tol=tol)
+        self.margin = margin
+
+    def generate_transform_matrix(self, n_features: int) -> np.ndarray:
+        D = np.eye(n_features, k=-1) + np.eye(n_features, k=1) - 2 * np.eye(n_features)
+        D[0:self.margin, 0] = -2
+        D[0:self.margin, 1] = 2
+        D[-1:-self.margin, -1] = -2
+        D[-1:-self.margin, -2] = 2
```

Comment:
I don't know the exact meaning of `margin` but if you want to make the following matrix

```
>>> D = np.eye(6, k=-1) + np.eye(6, k=1) - 2 * np.eye(6)
>>> D
array([[-2.,  1.,  0.,  0.,  0.,  0.],
       [ 1., -2.,  1.,  0.,  0.,  0.],
       [ 0.,  1., -2.,  1.,  0.,  0.],
       [ 0.,  0.,  1., -2.,  1.,  0.],
       [ 0.,  0.,  0.,  1., -2.,  1.],
       [ 0.,  0.,  0.,  0.,  1., -2.]])
```

to ( with margin = 4 for instance )

```
>>> D
array([[-2.,  2.,  0.,  0.,  0.,  0.],
       [-2.,  2.,  1.,  0.,  0.,  0.],
       [-2.,  2., -2.,  1.,  2., -2.],
       [-2.,  2.,  1., -2.,  2., -2.],
       [ 0.,  0.,  0.,  1.,  2., -2.],
       [ 0.,  0.,  0.,  0.,  2., -2.]])
```

then, the last 2 lines should be

```
D[-self.margin:, -1] = -2
D[-self.margin:, -2] = 2
```

`D[-1:-4, -1]` or `D[-1:-4, -2]` doesn't change anything.

Plus, in my understanding, the default value of margin should be zero than 1. What do you think about this? @sutkss

Reply:
@tksmd I realized that we don't need the parameter `margin`.
I modified the function generate_transform_matrix in TrendFilteringADMM, and I added some test case to check generate_transform_matrix function.